File: rich_game_mechnism/src/com/tw/core/places/Estate.java
package com.tw.core.places;

import com.tw.core.Player;
import com.tw.core.commands.CommandFactory;

/**
 * Created by pzzheng on 11/16/16.
 */
public class Estate extends Place {
    private int emptyPrice;
    private Player owner;
    private Level level;

    public Estate(int emptyPrice) {
        this.emptyPrice = emptyPrice;
        level = Level.EMPTY;
    }

    public int getEmptyPrice() {
        return emptyPrice;
    }

    @Override
    public Player.Status comeHere(Player player) {
        player.moveTo(this);
        return estateType(player).action(player, this);
    }

    public Type estateType(Player player) {
        if (owner == null) return Type.EMPTY;
        if (owner.equals(player)) return Type.OWNER;
        return Type.OTHER;
    }

    public void setOwner(Player player) {
        owner = player;
    }

    public Level getLevel() {
        return level;
    }

    public void upgrade() {
        level = Level.values()[level.ordinal() + 1];
    }

    public Player getOwner() {
        return owner;
    }

    public void beEmpty() {
        setOwner(null);
        level = Level.EMPTY;
    }

    public enum Type {
        EMPTY {
            @Override
            Player.Status action(Player player, Estate estate) {
                if (player.getAsests().getFunds() < estate.emptyPrice) return player.endTurn();
                player.setLastCommand(CommandFactory.BuyEstate(estate));
                return player.waitForResponse();
            }
        }, OTHER {
            @Override
            Player.Status action(Player player, Estate estate) {
                if (!player.isLucky() && !estate.getOwner().isStucked()) {
                    int charge = (int) (1.0 * estate.getEmptyPrice() * (estate.getLevel().ordinal() + 1) / 2);
                    player.getAsests().addFund(-charge);
                    if (player.getAsests().getFunds() < 0) return player.bankrupt();
                }
                return player.endTurn();
            }
        }, OWNER {
            @Override
            Player.Status action(Player player, Estate estate) {
                if (player.getAsests().getFunds() < estate.emptyPrice) return player.endTurn();
                player.setLastCommand(CommandFactory.UpgradeEstate(estate));
                return player.waitForResponse();
            }
        };

        abstract Player.Status action(Player player, Estate estate);
    }

    public enum Level {EMPTY, THATCH, FOREIGN_STYLE, SKYSCRAPER}
}

File: ruby/rich_game_ruby/src/places/estate.rb
require_relative 'place'
class Estate < Place
  attr_accessor :owner
  attr_reader :emptyPrice, :level

  def initialize(emptyPrice)
    @emptyPrice = emptyPrice
    @owner = nil
    @level = Level::EMPTY
  end

  def comeHere(player)
    player.moveTo(self)
    typeFor(player).action(player, self)
  end

  def typeFor(player)
    return Type::EMPTY if (@owner == nil)
    return Type::OWNER if (@owner == player)
    return Type::OTHER if (@owner != player)
  end

  def toll
    (@emptyPrice * (@level+1) / 2.0).to_i
  end

  def upgrade
    @level += 1
  end

  class Type
    EMPTY = Type.new
    def EMPTY.action(player, estate)
      return player.endTurn if (player.asset.fund < estate.emptyPrice)
      player.lastResponsiveCommand = CommandFactory.BuyEstate(estate)
      return player.waitForResponse
    end

    OWNER = Type.new
    def OWNER.action(player, estate)
      return player.endTurn if (player.asset.fund < estate.emptyPrice || estate.level == Estate::Level::SKYSCRAPER)
      player.lastResponsiveCommand = CommandFactory.UpgradeEstate(estate)
      return player.waitForResponse
    end

    OTHER = Type.new
    def OTHER.action(player, estate)
      return player.endTurn if (player.isLucky || estate.owner.isStucked)
      return player.bankrupt if (player.asset.fund < estate.toll)
      player.asset.chargeToll(estate)
      estate.owner.asset.earnToll(estate)
      return player.endTurn
    end

    def action(player, estate)
    end

  end

  module Level
    EMPTY = 0
    THATCH = 1
    FOREIGN_STYLE = 2
    SKYSCRAPER = 3
  end
end